move elasticity to rheology model

jdannberg opened the pull request:
This is a suggestion for moving the elasticity to a new rheology class. Although this looks like a lot of lines, most of these lines are just moved from viscoelastic to the new file. It would be great if I could get some feedback on the functions, if they make sense like this, or if they should be different/have different arguments. 

In particular, the `fill_elastic_force_outputs` and `fill_reaction_outputs` right now simply get the material model outputs and use some of these values as inputs (the viscosity), but fill other properties (force outputs, reaction outputs).

gassmoeller commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -0,0 +1,159 @@
+          get_elastic_shear_moduli () const;
+
+          /**
+           * Given the (viscous) viscosity and the shear modulus, compute the viscoelastic
```
```suggestion
           * Given the (viscous or visco-plastic) viscosity and the shear modulus, compute the viscoelastic
```

jdannberg commented:
I added the viscoelasto-plastic model, so it would be great to get a more thorough review, @gassmoeller. 

Also, @naliboff, these material model used different averaging schemes! Please tell me which one is the correct one (averaging elastic moduli, or averaging viscoelastic viscosities) and then I will implement this once and for all and remove the other option.

naliboff commented:
@jdannberg - The overall structure looks really good! More comments later today.

Regarding the averaging, after the discussion with @bobmyhill I think the best approach is to average the elastic shear moduli before calculating the viscoelastic viscosities (e.g, method in viscoelastic material model).  Would you like me to do a PR for this fix or implement it here?

jdannberg commented:
@naliboff:

> I think the best approach is to average the elastic shear moduli before calculating the viscoelastic viscosities

That's what I do in this pull request. So there's no need to change anything :smiley:

naliboff commented:
Then lets stick with the averaging method you have implemented 👍 Unlikely anyone has started using the VEP model for a production run since it was merged, so no reason not to change it now.

gassmoeller commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -0,0 +1,381 @@
+        // We also use this parameter when we are still *before* the first time step,
+        // i.e., if the time step number is numbers::invalid_unsigned_int.
+        const double dte = ( ( this->get_timestep_number() > 0 &&
+                               this->get_timestep_number() != numbers::invalid_unsigned_int &&
```
This should be equivalent to 

```suggestion
                               this->simulator_is_past_initialization() &&
```

tjhei commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -0,0 +1,159 @@
+           */
+          void
+          fill_elastic_force_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
+                                      const std::vector<double> average_elastic_shear_moduli,
```
add reference

tjhei commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -0,0 +1,159 @@
+       * MaterialModel::Interface::evaluate() function.
+       */
+      template <int dim>
+      class ElasticAdditionalOutputs : public NamedAdditionalMaterialOutputs<dim>
```
Is putting this into the Rheology namespace intentional?

jdannberg replied:
Well I thought elastic outputs are something that's really part of the rheology model, so I just put it into that namespace. What would be a reason for moving it out of the namespace?

tjhei replied:
The reason I mention this is that you are moving this class from the MaterialModel namespace. I am not sure if others are using this class, but you would break those plugins.

jdannberg replied:
I checked, an the functionality was added more than a year ago (#1593). I'll move this out of the namespace.

tjhei commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -0,0 +1,159 @@
+      class ElasticAdditionalOutputs : public NamedAdditionalMaterialOutputs<dim>
+      {
+        public:
+          ElasticAdditionalOutputs(const unsigned int n_points);
```
```suggestion
          explicit ElasticAdditionalOutputs(const unsigned int n_points);
```

tjhei commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -0,0 +1,159 @@
+           */
+          void
+          fill_reaction_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
+                                 const std::vector<double> average_elastic_shear_moduli,
```
same here

tjhei commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,16 @@ namespace aspect
 
 
+      /**
+       * Used for calculating average elastic shear modulus and viscosity
```
can you explain what it does?

jdannberg replied:
I was just moving this function, but after checking what it actually does, I found it was just duplicating code, so I removed it. That also solves your comment about the name below.

tjhei commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,16 @@ namespace aspect
+       * Used for calculating average elastic shear modulus and viscosity
+       */
+      template <int dim>
+      double calculate_average_vector (const std::vector<double> &composition,
```
it doesn't calculate an "average vector" but it averages a vector into a double, right? Can you think of a better name? calculate_average_of_vector? average_vector?
And sorry, I realize you are just moving things around.

tjhei commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -0,0 +1,381 @@
+      ElasticAdditionalOutputs<dim>::get_nth_output(const unsigned int idx) const
+      {
+        AssertIndexRange (idx, 1);
+        switch (idx)
```
you can delete the whole switch.

tjhei commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -0,0 +1,381 @@
+                                   "compositional field is called stress_yz."));
+          }
+        else
+          ExcNotImplemented();
```
that doesn't do what you want without the assert. ;-)

jdannberg replied:
Wow, I didn't even see that...

jdannberg commented:
I think I addressed all comments (and I also rebased, as there were conflicts with master). 
@tjhei and @gassmoeller: Thank you for your reviews. Next time I will read the code that I am moving more carefully! You caught lots of things that didn't make any sense.

tjhei commented:
> Next time I will read the code that I am moving more carefully!

sorry for nitpicking code that you just wanted to move...

jdannberg commented:
> sorry for nitpicking code that you just wanted to move...

I know I usually complain about this, but this time I feel it was totally justified. The whole pull request was about cleaning up the material models, and this actually saved us 50 lines of code and made everything much more readable.

tjhei commented:
Looks good, but I still think we should keep the AdditionalOutputs in the same namespace.

jdannberg commented:
Yes, I agree. Done.
